src/Functions/parseDateTime.cpp

```diff
@@ -0,0 +1,1747 @@
+                    if (std::string_view(cur, literal.size()) != literal)
+                        throw Exception(
+                            ErrorCodes::LOGICAL_ERROR, "Expect literal {} but {} provided", literal, std::string_view(cur, literal.size()));
+                    cur += literal.size();
+                    return cur;
+                }
+            }
+
+            template <typename T, bool check_space = true>
+            static Pos readNumber2(Pos cur, Pos end, T & res)
+            {
+                if constexpr (check_space)
+                    ensureSpace(cur, end, 2, "readNumber2 requires size >= 2");
+
+                res = (*cur - '0');
+                ++cur;
+                res = res * 10 + (*cur - '0');
+                ++cur;
+                return cur;
+            }
+
+            template <typename T, bool check_space = true>
+            static Pos readNumber3(Pos cur, Pos end, T & res)
+            {
+                if constexpr (check_space)
+                    ensureSpace(cur, end, 3, "readNumber4 requires size >= 3");
+
+                res = (*cur - '0');
+                ++cur;
+                res = res * 10 + (*cur - '0');
+                ++cur;
+                res = res * 10 + (*cur - '0');
+                ++cur;
+                return cur;
+            }
+
+            template <typename T, bool check_space = true>
+            static Pos readNumber4(Pos cur, Pos end, T & res)
+            {
+                if constexpr (check_space)
+                    ensureSpace(cur, end, 4, "readNumber4 requires size >= 4");
+
+                res = (*cur - '0');
+                ++cur;
+                res = res * 10 + (*cur - '0');
+                ++cur;
+                res = res * 10 + (*cur - '0');
+                ++cur;
+                res = res * 10 + (*cur - '0');
+                ++cur;
+                return cur;
+            }
+
+            static void ensureSpace(Pos cur, Pos end, size_t len, const String & msg)
+            {
+                if (cur > end || cur + len > end)
+                    throw Exception(ErrorCodes::LOGICAL_ERROR, "Unable to parse because {}", msg);
+            }
+
+            template <bool check_space = true>
+            static Pos assertChar(Pos cur, Pos end, char ch)
+            {
+                if constexpr (check_space)
+                    ensureSpace(cur, end, 1, "assertChar requires size >= 1");
+
+                if (*cur != ch)
+                    throw Exception(ErrorCodes::LOGICAL_ERROR, "Expect char {}, but {} provided", String(ch, 1), String(*cur, 1));
+
+                ++cur;
+                return cur;
+            }
+
+            static Pos mysqlDayOfWeekTextShort(Pos cur, Pos end, Date & date)
+            {
+                ensureSpace(cur, end, 3, "Parsing DayOfWeekTextShort requires size >= 3");
+
+                String text(cur, 3);
+                Poco::toLowerInPlace(text);
```

Comment:
If there is something from boost, we should prefer it over poco (which is less popular). --> `boost::to_lower()`

By the way, why are we case-insensitive here? It is IMHO totally okay to expect the same string which is produced by `formatDateTime(InJodaSyntax)()`. Perhaps case-insensitive matching could be made a new function (--> separate PR). If we are by default case-sensitive, we would not only not need lowercasing (l. (723), we could probably also use a `std::string_view` instead of a `String` in l. 722 and thus save a heap allocation.

Reply:
boost::to_lower is used now.

Reply:
About the case-sensitive or case-insensitive issue: I noticed that text parsing in similar spark to_timestamp function is case-insensitive, which I think is more friendly to users.

Reply:
The examples in spark: 
``` sql 
+--------------------------------------------------+
| to_timestamp(2020-01-01 07 am, yyyy-MM-dd HH a)  |
+--------------------------------------------------+
| 2020-01-01 07:00:00.0                            |


+--------------------------------------------------+
| to_timestamp(2020-01-01 07 AM, yyyy-MM-dd HH a)  |
+--------------------------------------------------+
| 2020-01-01 07:00:00.0                            |


```

Reply:
Alright, thanks.

Just for my understanding: does case-insensitivity also work if the output of lowercase and uppercase formatters is actually different? E.g. `%M/%m` or `%w/%w` in mySQL syntax or `E/e` in joda syntax? Or is case-insensitivity only applied for cases where there is no difference?